lab_1_keywords_tfidf/main.py

```diff
@@ -3,103 +3,157 @@
 Extract keywords based on frequency related metrics
 """
 from typing import Optional, Union
+import math
+
+
+def check_input_type(check_arg, check_type, check_token=None, check_value=None, can_be_empty=True):
+    if not isinstance(check_arg, check_type):
+        return False
+    if not check_arg and not can_be_empty:
+        return False
+    if check_type == list:
+        for i in check_arg:
+            if not isinstance(i, check_token):
+                return False
+    if check_type == dict:
+        for i in check_arg:
+            if not isinstance(i, check_value):
+                return False
+    return True
+
+
+def check_num(num_arg):
+    if not isinstance(num_arg, int) or type(num_arg) == bool:
+        return False
+    if num_arg < 0:
+        return False
+    return True
 
 
 def clean_and_tokenize(text: str) -> Optional[list[str]]:
     """
     Removes punctuation, casts to lowercase, splits into tokens
-
     Parameters:
     text (str): Original text
-
     Returns:
     list[str]: A sequence of lowercase tokens with no punctuation
-
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if not check_input_type(text, str):
+        return None
+    text_small = text.lower().strip()
+    tokens_str = ''
+    for i in text_small:
+        if i.isalnum() is True or i == ' ' or i == '\n':
+            tokens_str += i
+    tokens = tokens_str.split()
+    return tokens
 
 
 def remove_stop_words(tokens: list[str], stop_words: list[str]) -> Optional[list[str]]:
     """
     Excludes stop words from the token sequence
-
     Parameters:
     tokens (List[str]): Original token sequence
     stop_words (List[str]: Tokens to exclude
-
     Returns:
     List[str]: Token sequence that does not include stop words
-
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if not check_input_type(tokens, list, str, False) or \
+            not check_input_type(stop_words, list, str):
+        return None
+    no_stop_words = []
+    for x in tokens:
+        if x not in stop_words:
+            no_stop_words.append(x)
+        else:
+            continue
```

Comment:
is the `else` branch necessary here? can you explain what `continue` does?

Reply:
forgot to delete, fixed